### ISSUE TITLE
Callback protocols with more than just a `__call__` attribute report an error

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2560]
pyright version `1.1.183` seems to have introduced a regression when using callback protocols that have more than just a `__call__` attribute:
```py
from typing import Protocol


class SomeProtocol(Protocol):
    __name__: str

    def __call__(self) -> str:
        ...


def some_wrapper(func: SomeProtocol):
    print(func.__name__)


@some_wrapper
# Argument of type "() -> str" cannot be assigned to parameter "func" of type "SomeProtocol" in function "some_wrapper"
#  "object" is incompatible with protocol "SomeProtocol"
#    "__name__" is not present
#    "__call__" is not present
def some_func() -> str:
    ...

```

This issue came up in the past here: https://github.com/microsoft/pyright/issues/2179, where it was decided that callback protocols would support multiple attributes. Has the stance changed since then, or is this a bug?

Above code type checks without errors in version `1.1.182`, but versions 1.1.183+ report the error seen above.

**Comment by erictraut** [comment]
Thanks for reporting the problem (again).

The fix for #2179 introduced a type validation hole, which I fixed in this [this change](https://github.com/microsoft/pyright/issues/2512) as part of 1.1.183. As part of this subsequent change, I special-cased fields that are supported by all functions, but this special-casing didn't include the `__name__` field. I've fixed that omission and added a more thorough test case to prevent future regressions in this code.
